backend.go

```diff
@@ -961,48 +961,129 @@ func (backend *Backend) DecodeMatrix(frags []ValidatedFragment, pieceSize int) (
-	FragRangeEnd      int // End offset in each K+M fragments
-	DecodedRangeStart int // Start offset in decoded data
-	DecodedRangeEnd   int // end offset in decoded data
-}
+	ReqStartIncl int
+	ReqEndIncl   int
 
-// GetRangeMatrix returns the bounds of each data fragments to get to satisfy
-// {start, end} range
-func (backend *Backend) GetRangeMatrix(start, end, chunksize, fragSize int) *RangeMatrix {
-	blockSize := chunksize
-	groupSize := blockSize * backend.K
+	/* The fragments that the range spans.  */
+	FragFirstIncl int
+	FragCount     int
 
-	// check that range can be satisfied
-	nrChunkByFrag := fragSize / (backend.headerSize + chunksize)
-	trueFragLen := nrChunkByFrag * chunksize
-	linearizedDataLen := trueFragLen * backend.K
+	/* The range in each fragment to be queried satisfy the requested range. */
+	InFragRangeStartIncl int
+	InFragRangeEndExcl   int
+
+	/* The requested range relative to the decoded buffer. */
+	LinearizedRangeStartIncl int
+	DecodedRangeStartIncl    int
+}
 
-	if start > linearizedDataLen || end > linearizedDataLen || start > end {
+/*
+ * Returns the ranges to read a matrix encoded data. This function tries
+ * to minimize the number of request to perform depending on the requested
+ * range.
+ *
+ * There are a few design choices that make the result not always obvious.
+ * 		1. Each fragment range is always identical.
+ *      2. When the requested range wraps around fragments all fragments
+ *         are always queried.
+ *
+ * (1) is currently necessary to avoid querying multiple time the same
+ * fragment in case of failures (to reconstruct the data). This prefer
+ * performing the minimum amount of IO requests, instead of reading the minimum
+ * amount of data. We could lift this constraint if the caller would stream
+ * group at a time but that would require the backend to have matching
+ * alignement constraints.
+ *
+ * (2) could also be lifted. This is currently done to avoid changing the way
+ * the current decoding is performed. To work, it currently requires consecutive
+ * fragments. We then can't leave gaps in fragments. For example if
+ * we take a erasure code with 4 data fragments of 4 chunks,
+ * with the requested range represented by a '*' and the resulting fragment
+ * ranges by '[]':
+ *
+ *     p1 [-[- *]-]    The request here start at the end of the 2nd chunk in p4
+ *        [-[- *]-]    then wraps around in the subsequent chunks in p1 and p2.
+ *     .. [-[- -]-]
+ *     p4 [-[* -]-]    When this occurs, we will still query, p4 just to decode
+ *                     the relevant requested range. The decoded buffer
+ *                     will look like [p1 p2 p3 p4(*) p1(*) p2(*) p3 p4].
+ *                     Chunks not marked with a '*' are discarded. Note how
+ *                     the heading and trailing is unecessary and could be
+ *                     discarded in the ideal case.
+ *
+ * Perfect cases occur when the request span a single group (column):
```

Comment:
IIUC We dont have to load the unecessary p1 in this case since no gaps between p2->p4?

Reply:
Yes